### PR TITLE
Let the warehouse use the rack-maps for speed

### PR DESCRIPTION
**Pull request by Mekle001**
# Changes proposed in this pull request:
- Modify the warehouse resolver to use a fast path for single stack requests

Review please

### CODE REVIEW

**Comment by Raycoms** [comment]
:shipit: 

[Approved with PullApprove](https://pullapprove.com/ldtteam/minecolonies/pull-request/5391/?utm_source=github-pr&utm_medium=comment-badge&utm_campaign=ldtteam/minecolonies)

File: src/api/java/com/minecolonies/api/tileentities/TileEntityRack.java [hunk 1]
```diff
@@ -133,15 +133,7 @@ public int getFreeSlots()
...
-        for (final Map.Entry<ItemStorage, Integer> entry : content.entrySet())
-        {
-            if (compareStorage.equals(entry.getKey()))
-            {
-                return true;
-            }
-        }
-        return false;
+        return content.get(new ItemStorage(stack, ignoreDamageValue)) != null;
     }
```
Comment (Raycoms) [inline src/api/java/com/minecolonies/api/tileentities/TileEntityRack.java:141]: lol, easy one

**Review by someaddons — DISMISSED** [review]

File: src/api/java/com/minecolonies/api/tileentities/TileEntityRack.java (hunk 1 above)
Comment (someaddons) [reply src/api/java/com/minecolonies/api/tileentities/TileEntityRack.java:141]: oh actually != isnt needed, could just return content.contains(...)

Comment (Mekle001) [reply src/api/java/com/minecolonies/api/tileentities/TileEntityRack.java:141]: Fair enough. :)

**Review by marchermans — COMMENTED** [review]
Not really a good solution.
PR put on hold.

File: src/api/java/com/minecolonies/api/colony/requestsystem/requestable/Stack.java [hunk 2]
```diff
@@ -73,8 +77,7 @@ public Stack(@NotNull final ItemStack stack)
      * @param count    the count.
      * @param minCount the min count.
      */
-    public Stack(@NotNull final ItemStack stack, final int count, final int minCount)
-    {
```
Comment (Raycoms) [inline src/api/java/com/minecolonies/api/colony/requestsystem/requestable/Stack.java]: the formatting in the file got messed up =D

File: src/api/java/com/minecolonies/api/colony/requestsystem/requestable/StackList.java [hunk 3]
```diff
@@ -375,4 +375,9 @@ public String getDescription()
     {
         return description;
     }
+
+    @Override
```
Comment (Raycoms) [inline src/api/java/com/minecolonies/api/colony/requestsystem/requestable/StackList.java:379]: also formatting

File: src/main/java/com/minecolonies/coremod/colony/requestsystem/resolvers/WarehouseConcreteRequestResolver.java [hunk 4]
```diff
@@ -0,0 +1,280 @@
...
+import com.minecolonies.api.colony.requestsystem.requestable.IDeliverable;
+import com.minecolonies.api.colony.requestsystem.requestable.deliveryman.Delivery;
+import com.minecolonies.api.colony.requestsystem.requester.IRequester;
+import com.minecolonies.api.colony.requestsystem.token.IToken;
+import com.minecolonies.api.util.ItemStackUtils;
+import com.minecolonies.api.util.Log;
+import com.minecolonies.api.util.constant.TranslationConstants;
+import com.minecolonies.api.util.constant.TypeConstants;
+import com.minecolonies.coremod.colony.Colony;
+import com.minecolonies.coremod.colony.buildings.workerbuildings.BuildingWareHouse;
```
Comment (Raycoms) [inline src/main/java/com/minecolonies/coremod/colony/requestsystem/resolvers/WarehouseConcreteRequestResolver.java]: We should have an AbstractWareHouseRequestResolver that has the parts both have in common

**Review by Raycoms — CHANGES_REQUESTED** [review]
Besides what I've pointed out inline, we will need also an update handler again to reset the RS and load the new resolver to all warehouses.

File: src/api/java/com/minecolonies/api/colony/requestsystem/requestable/IConcreteDeliverable.java [hunk 5]
```diff
@@ -0,0 +1,17 @@
+package com.minecolonies.api.colony.requestsystem.requestable;
+
+import net.minecraft.item.ItemStack;
+
+import java.util.List;
+
+/**
+ * An {@link IConcreteDeliverable} is an Requestable that can be looked up fast in the warehouse, and delivered.
+ */
+public interface IConcreteDeliverable extends IDeliverable
```
Comment (marchermans) [inline src/api/java/com/minecolonies/api/colony/requestsystem/requestable/IConcreteDeliverable.java:10]: Override the matches function from IDeliverable and make it the default implementation

File: src/main/java/com/minecolonies/coremod/colony/requestsystem/resolvers/WarehouseConcreteRequestResolver.java (hunk 4 above)
Comment (Mekle001) [reply src/main/java/com/minecolonies/coremod/colony/requestsystem/resolvers/WarehouseConcreteRequestResolver.java]: Done

File: src/main/java/com/minecolonies/coremod/colony/requestsystem/resolvers/WarehouseConcreteRequestResolver.java [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+            {
+                Log.getLogger().error(e);
+            }
+        }
+
+        return false;
+    }
+
+    @Override
+    protected boolean isRequestFromSelf(final IRequest<?> requestToCheck)
```
Comment (Raycoms) [inline src/main/java/com/minecolonies/coremod/colony/requestsystem/resolvers/WarehouseConcreteRequestResolver.java]: Couldn't this and the getPriority go into the abstract?

Comment (Mekle001) [reply src/main/java/com/minecolonies/coremod/colony/requestsystem/resolvers/WarehouseConcreteRequestResolver.java]: The getPriority can. I guess we don't have a scenario where it would be valid to come from one of the warehouse resolvers to the other, so yeah, this can probably move too.

File: src/main/java/com/minecolonies/coremod/colony/requestsystem/resolvers/WarehouseRequestResolver.java [hunk 7]
```diff
@@ -48,222 +32,20 @@ public WarehouseRequestResolver(
...
-    {
-        return new TranslationTextComponent(TranslationConstants.COM_MINECOLONIES_BUILDING_WAREHOUSE_NAME);
+        return wareHouses.stream()
+                .anyMatch(wareHouse -> wareHouse.hasMatchingItemStackInWarehouse(itemStack -> requestToCheck.getRequest().matches(itemStack),
+                requestToCheck.getRequest().getMinimumCount()));
     }
 
     @Override
-    public int getPriority()
+    protected AbstractWarehouseRequestResolver newInstanceOfSelf(@NotNull final ILocation location, @NotNull final IToken<?> token)
```
Comment (marchermans) [inline src/main/java/com/minecolonies/coremod/colony/requestsystem/resolvers/WarehouseRequestResolver.java]: No this does not work.
This is completely hats on ass logic.

Why is this here!?!?!?

File: src/main/java/com/minecolonies/coremod/colony/requestsystem/resolvers/core/AbstractWarehouseRequestResolver.java [hunk 8]
```diff
@@ -0,0 +1,285 @@
...
+        if (requestToCheck.getRequester().getClass().equals(this.getClass()))
+        {
+            return true;
+        }
+        return false;
+    }
+    /**
+     * Let the decendant class create an instance for use
+     */
+    protected abstract AbstractWarehouseRequestResolver newInstanceOfSelf(@NotNull final ILocation location, @NotNull final IToken<?> token);
```
Comment (marchermans) [inline src/main/java/com/minecolonies/coremod/colony/requestsystem/resolvers/core/AbstractWarehouseRequestResolver.java]: WTF is this doing here?!?!?!?

File: src/main/java/com/minecolonies/coremod/colony/requestsystem/resolvers/WarehouseConcreteRequestResolver.java [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+import org.jetbrains.annotations.NotNull;
+
+import java.util.Objects;
+import java.util.Optional;
+import java.util.Set;
+
+/**
+ * ----------------------- Not Documented Object ---------------------
+ */
+public class WarehouseConcreteRequestResolver extends AbstractWarehouseRequestResolver
```
Comment (marchermans) [inline src/main/java/com/minecolonies/coremod/colony/requestsystem/resolvers/WarehouseConcreteRequestResolver.java:18]: The factory for this class is missing.

**Review by marchermans — CHANGES_REQUESTED** [review]
The formatting is a mess.
The PR is half done, half not done.

Stuff is undocumented everywhere.
And we are still only doing this half assed. Either make the resolver for the warehouse always use the ItemStorage variant, or never. But not this half way.

File: src/main/java/com/minecolonies/coremod/colony/requestsystem/resolvers/WarehouseConcreteRequestResolver.java (hunk 9 above)
Comment (Mekle001) [reply src/main/java/com/minecolonies/coremod/colony/requestsystem/resolvers/WarehouseConcreteRequestResolver.java:18]: And, this is why the code you went "wtf" about. I didn't realize that there was a factory pattern set up for the WarehouseRequestResolver, as it doesn't actually appear to be in use.  I'll gladly add the factory and remove the  newInstanceOfSelf.

Comment (marchermans) [reply src/main/java/com/minecolonies/coremod/colony/requestsystem/resolvers/WarehouseConcreteRequestResolver.java:18]: The entire RS uses factories that have to be registered. They handle the entire IO for it.

Comment (Mekle001) [reply src/main/java/com/minecolonies/coremod/colony/requestsystem/resolvers/WarehouseConcreteRequestResolver.java:18]: So, apparently I figured it out, added it, and forgot. The factories actually are there, but with the formatting issue in Stack.java it was not apparent.

File: src/api/java/com/minecolonies/api/colony/requestsystem/requestable/IConcreteDeliverable.java (hunk 5 above)
Comment (Mekle001) [reply src/api/java/com/minecolonies/api/colony/requestsystem/requestable/IConcreteDeliverable.java:10]: Let's discuss this. I started to do that and it felt inside out, because we're comparing two collections to see if there is any overlap. The deliverable list is likely the smallest of those, and thus needs to be the one we iterate over. I'm guessing the matches you're proposing would be to send in a list, but that would be the full inventory and negate the gains we're looking for.

File: src/main/java/com/minecolonies/coremod/colony/requestsystem/resolvers/core/AbstractWarehouseRequestResolver.java (hunk 8 above)
Comment (Mekle001) [reply src/main/java/com/minecolonies/coremod/colony/requestsystem/resolvers/core/AbstractWarehouseRequestResolver.java]: This is gone.

File: src/main/java/com/minecolonies/coremod/colony/requestsystem/resolvers/WarehouseRequestResolver.java (hunk 7 above)
Comment (Mekle001) [reply src/main/java/com/minecolonies/coremod/colony/requestsystem/resolvers/WarehouseRequestResolver.java]: This is gone.

**Review by marchermans — APPROVED** [review]

**Review by someaddons — APPROVED** [review]
